File: 0446-arithmetic-slices-ii-subsequence/0446-arithmetic-slices-ii-subsequence.py
# LarrryNY's Code

class Solution:
    def numberOfArithmeticSlices(self, nums: List[int]) -> int:
        N = len(nums)
    
        lookup = []
        ans = 0

        for i in range(N):
            lookup.append(collections.defaultdict(int))

            for j in range(i):
                delta = nums[i] - nums[j]

                if delta in lookup[j]:
                    count = lookup[j][delta]
                    ans += count
                    lookup[i][delta] += count + 1
                else:
                    lookup[i][delta] += 1

        return ans        

File: 0446-arithmetic-slices-ii-subsequence/0446-arithmetic-slices-ii-subsequence.cpp
class Solution {
public:
    int numberOfArithmeticSlices(vector<int>& nums) {
        long long n,i,j,commonDiff,ans;
        n = nums.size();
                                        
        unordered_map<long long,long long> M[n]; // M[i]: {commonDiff,cnt of AP's len >= 2 ending with nums[i]} 
        
        ans = 0;
        for(i=0 ; i<n ; i++){
            for(j=0 ; j<=(i-1) ; j++){
                commonDiff = (long long)(nums[i]) - (long long)(nums[j]);
                if(M[j].find(commonDiff) != M[j].end()){
                    ans += M[j][commonDiff];
                    M[i][commonDiff] += (M[j][commonDiff] + 1);
                }
                else{
                    M[i][commonDiff] += 1;
                }
            }
        }
        
        return ans;
    }
};